README.md
# LoadingHud

[![pub package](https://img.shields.io/pub/v/loading_hud.svg)](https://img.shields.io/pub/v/loading_hud)

A dialog with a loading indicator.

![showcase](screenshots/screen_shot.gif)

## Getting Started

### Add dependency

```yaml
dependencies:
  loading_hud: ^0.2.0
```

### Examples

```dart
var loadingHud = LoadingHud(
  context,
  cancelable: true,                  // Cancelable when pressing Android back key
  canceledOnTouchOutside: true,      // Cancelable when touch outside of the LoadingHud
  dimBackground: true,               // Dimming background when LoadingHud is showing
  hudColor: Color(0x99000000),       // Color of the ProgressHud
  indicator: DefaultLoadingIndicator(
    color: Colors.white,
  ),
  iconSuccess: Icon(                 // Success icon
    Icons.done,
    color: Colors.white,
  ),
  iconError: Icon(                   // Error icon
    Icons.error,
    color: Colors.white,
  ),
  future: loadingSomething();
);

Future<Text> loadingSomething() async {
  try {
    var data = await loadingSomthingTimeConsuming();
    return Text('That was great!');
  } catch (e) {
    throw Text('Something went wrong.');
  }
}
  
/// Show LoadingHud
loadingHud.show();

/// Dismiss LoadingHud
loadingHud.dismiss();
```



CHANGELOG.md
## 0.1.0

- Initial release.

## 0.2.0

- Support displaying success and error message when loading is completed.


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:progresshud/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/loading_hud.dart
import 'dart:async';
import 'package:flutter/widgets.dart';
import 'package:loading_hud/loading_indicator.dart';

class LoadingHud<T extends Text> {
  final BuildContext context;
  final bool cancelable;
  final bool canceledOnTouchOutside;
  final bool dimBackground;
  final Color hudColor;
  final Widget indicator;
  final Widget iconSuccess;
  final Widget iconError;
  final Future<T> future;
  final Duration autoDismissDuration;

  LoadingHud(
    this.context, {
    this.cancelable = true,
    this.canceledOnTouchOutside = true,
    this.dimBackground = true,
    this.hudColor = const Color(0xDDFFFFFF),
    this.indicator = const DefaultLoadingIndicator(),
    this.iconSuccess,
    this.iconError,
    this.future,
    this.autoDismissDuration,
  });

  void show() {
    showGeneralDialog(
      context: context,
      pageBuilder: (BuildContext context, Animation<double> animation,
          Animation<double> secondaryAnimation) {
        return WillPopScope(
          child: _HudContent<T>(
            hudColor: hudColor,
            indicator: indicator,
            iconSuccess: iconSuccess,
            iconError: iconError,
            future: future,
            autoDismissDuration: autoDismissDuration,
          ),
          onWillPop: () {
            return Future.value(cancelable);
          },
        );
      },
      barrierColor: dimBackground ? Color(0x33000000) : null,
      barrierLabel: 'Dismiss',
      barrierDismissible: canceledOnTouchOutside,
      transitionDuration: const Duration(milliseconds: 250),
      useRootNavigator: true,
    );
  }

  void dismiss() {
    Navigator.of(context, rootNavigator: true).pop();
  }
}

class _HudContent<T extends Text> extends StatefulWidget {
  final Color hudColor;
  final Widget indicator;
  final Widget iconSuccess;
  final Widget iconError;
  final Future<T> future;
  final Duration autoDismissDuration;

  _HudContent({
    this.hudColor,
    this.indicator,
    this.iconSuccess,
    this.iconError,
    this.future,
    this.autoDismissDuration,
  });

  @override
  _HudContentState createState() => _HudContentState<T>();
}

class _HudContentState<T extends Text> extends State<_HudContent> {
  @override
  Widget build(BuildContext context) {
    return Center(
      child: FutureBuilder<T>(
        future: widget.future,
        builder: (BuildContext context, AsyncSnapshot<T> snapshot) {
          bool done = false;
          Widget widgetDone;
          if (snapshot.connectionState == ConnectionState.done) {
            if (snapshot.hasData) {
              done = true;
              widgetDone = _IconText(
                icon: widget.iconSuccess,
                text: snapshot.data,
              );
            } else if (snapshot.hasError) {
              done = true;
              widgetDone = _IconText(
                icon: widget.iconError,
                text: snapshot.error,
              );
            }
          }
          // Auto dismiss
          if (done && widget.autoDismissDuration != null) {
            Future.delayed(widget.autoDismissDuration, () {
              Navigator.of(context, rootNavigator: true).pop();
            });
          }
          return DecoratedBox(
            child: Padding(
              padding: const EdgeInsets.all(20),
              child: done
                  ? ConstrainedBox(
                      child: widgetDone,
                      constraints: BoxConstraints(
                        minHeight: 100,
                        minWidth: 100,
                      ),
                    )
                  : Padding(
                      padding: const EdgeInsets.all(35),
                      child: widget.indicator,
                    ),
            ),
            decoration: BoxDecoration(
              borderRadius: BorderRadius.all(
                Radius.circular(12),
              ),
              color: widget.hudColor,
            ),
          );
        },
      ),
    );
  }
}

class _IconText extends StatelessWidget {
  final Widget icon;
  final Text text;

  _IconText({
    @required this.icon,
    this.text,
  }) : assert(icon != null);

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.center,
      mainAxisAlignment: MainAxisAlignment.center,
      mainAxisSize: MainAxisSize.min,
      children: <Widget>[
        icon,
        Padding(
          padding: const EdgeInsets.only(top: 20),
          child: Text(
            text.data,
            style: text.style ??
                TextStyle(
                  fontSize: 16,
                  color: Color(0xFF000000),
                  decoration: TextDecoration.none,
                ),
          ),
        ),
      ],
    );
  }
}


lib/loading_indicator.dart
import 'dart:async';
import 'dart:ui';
import 'dart:math' as Math;
import 'package:flutter/widgets.dart';

class DefaultLoadingIndicator extends StatefulWidget {
  final Color color;

  const DefaultLoadingIndicator({
    this.color = const Color(0x99000000),
  });

  @override
  _DefaultLoadingIndicatorState createState() =>
      _DefaultLoadingIndicatorState();
}

class _DefaultLoadingIndicatorState extends State<DefaultLoadingIndicator> {
  final leavesCount = 12;
  Timer timer;
  double rotateDegree = 0;
  Paint paint;
  final leaveColors = <Color>[];
  final Path path = Path();

  @override
  void initState() {
    super.initState();
    _initLeavesColors();
    _initPaint();
    timer = Timer.periodic(Duration(milliseconds: 50), (_) {
      setState(() {
        rotateDegree += 30;
        rotateDegree %= 360;
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      size: Size(30, 30),
      painter: _DefaultIndicatorPainter(
        paint,
        widget.color,
        leavesCount,
        leaveColors,
        path,
        rotateDegree,
      ),
    );
  }

  @override
  void dispose() {
    super.dispose();
    timer.cancel();
  }

  void _initPaint() {
    paint = Paint()
      ..isAntiAlias = true
      ..style = PaintingStyle.fill;
  }

  void _initLeavesColors() {
    Color startColor = widget.color;
    Color endColor = Color(widget.color.value & 0x00FFFFFF | 0x11000000);
    for (int i = 0; i < leavesCount; i++) {
      double fraction = 1 - i * 30 / 360;
      leaveColors.add(_evaluate(fraction, startColor, endColor));
    }
  }

  Color _evaluate(double fraction, Color startValue, Color endValue) {
    int startInt = startValue.value;
    double startA = ((startInt >> 24) & 0xff) / 255.0;
    double startR = ((startInt >> 16) & 0xff) / 255.0;
    double startG = ((startInt >> 8) & 0xff) / 255.0;
    double startB = (startInt & 0xff) / 255.0;
    int endInt = endValue.value;
    double endA = ((endInt >> 24) & 0xff) / 255.0;
    double endR = ((endInt >> 16) & 0xff) / 255.0;
    double endG = ((endInt >> 8) & 0xff) / 255.0;
    double endB = (endInt & 0xff) / 255.0;
    // convert from sRGB to linear
    startR = Math.pow(startR, 2.2);
    startG = Math.pow(startG, 2.2);
    startB = Math.pow(startB, 2.2);
    endR = Math.pow(endR, 2.2);
    endG = Math.pow(endG, 2.2);
    endB = Math.pow(endB, 2.2);
    // compute the interpolated color in linear space
    double a = startA + fraction * (endA - startA);
    double r = startR + fraction * (endR - startR);
    double g = startG + fraction * (endG - startG);
    double b = startB + fraction * (endB - startB);
    // convert back to sRGB in the [0..255] range
    a = a * 255.0;
    r = Math.pow(r, 1.0 / 2.2) * 255.0;
    g = Math.pow(g, 1.0 / 2.2) * 255.0;
    b = Math.pow(b, 1.0 / 2.2) * 255.0;
    return Color(
        a.round() << 24 | r.round() << 16 | g.round() << 8 | b.round());
  }
}

class _DefaultIndicatorPainter extends CustomPainter {
  final Paint p;
  final Color color;
  final leavesCount;
  final leaveColors;
  final Path path;
  final double rotateDegree;

  _DefaultIndicatorPainter(
    this.p,
    this.color,
    this.leavesCount,
    this.leaveColors,
    this.path,
    this.rotateDegree,
  );

  @override
  void paint(Canvas canvas, Size size) {
    _drawFlower(canvas, size);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }

  void _drawFlower(Canvas canvas, Size size) {
    canvas.translate(size.width / 2, size.width / 2);
    canvas.rotate(rotateDegree * Math.pi / 180);
    canvas.save();
    double leafWidth = size.width / 12;
    path.reset();
    Rect rect = Rect.fromLTRB(
        -leafWidth / 2, -size.width / 2, leafWidth / 2, -size.width / 4);
    RRect rRect = RRect.fromRectAndRadius(rect, Radius.circular(leafWidth / 2));
    path.addRRect(rRect);
    path.close();
    for (int i = 0; i < leavesCount; i++) {
      Color color = leaveColors[i];
      p.color = color;
      canvas.drawPath(path, p);
      canvas.rotate(-30 * Math.pi / 180);
    }
    canvas.restore();
  }
}


lib/flutter_progress_hud.dart
import 'dart:async';
import 'package:flutter/widgets.dart';
import 'dart:math' as Math;

class FlutterProgressHud {
  final BuildContext context;
  final bool cancelable;
  final bool canceledOnTouchOutside;
  final bool dimBackground;
  final Color hudColor;
  final Color indicatorColor;

  FlutterProgressHud(this.context, {
    this.cancelable,
    this.canceledOnTouchOutside = true,
    this.dimBackground = true,
    this.hudColor = const Color(0xDDFFFFFF),
    this.indicatorColor = const Color(0x99000000),
  });

  void show() {
    showGeneralDialog(
      context: context,
      pageBuilder: (BuildContext context, Animation<double> animation,
                    Animation<double> secondaryAnimation) {
        return WillPopScope(
          child: _HudContent(
            hudColor: hudColor,
            indicatorColor: indicatorColor,
          ),
          onWillPop: () {
            return Future.value(cancelable);
          },
        );
      },
      barrierColor: dimBackground ? Color(0x33000000) : null,
      barrierLabel: 'Dismiss',
      barrierDismissible: canceledOnTouchOutside,
      transitionDuration: const Duration(milliseconds: 250),
      useRootNavigator: true,
    );
  }

  void dismiss() {
    Navigator.of(context, rootNavigator: true).pop();
  }
}

class _HudContent extends StatefulWidget {
  final Color hudColor;
  final Color indicatorColor;

  _HudContent({
                this.hudColor = const Color(0xDDFFFFFF),
                this.indicatorColor = const Color(0x99000000),
              });

  @override
  _HudContentState createState() => _HudContentState();
}

class _HudContentState extends State<_HudContent> {
  @override
  Widget build(BuildContext context) {
    return Center(
      child: Container(
        width: 120,
        height: 120,
        alignment: Alignment.center,
        child: ProgressIndicator(widget.indicatorColor),
        decoration: BoxDecoration(
          borderRadius: BorderRadius.all(
            Radius.circular(12),
          ),
          color: widget.hudColor,
        ),
      ),
    );
  }
}

class ProgressIndicator extends StatefulWidget {
  final Paint p;
  final Color color;
  final leavesCount = 12;
  final leaveColors = <Color>[];
  final Path path = Path();

  ProgressIndicator(this.color,) : p = Paint()
    ..isAntiAlias = true
    ..style = PaintingStyle.fill;

  @override
  _ProgressIndicatorState createState() => _ProgressIndicatorState();
}

class _ProgressIndicatorState extends State<ProgressIndicator> {
  Timer timer;
  double rotateDegree = 0;

  @override
  void initState() {
    super.initState();
    _initLeavesColors();
    timer = Timer.periodic(Duration(milliseconds: 50), (_) {
      setState(() {
        rotateDegree += 30;
        rotateDegree %= 360;
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      size: Size(30, 30),
      painter: _IndicatorPainter(widget.p, widget.color, widget.leavesCount,
          widget.leaveColors, widget.path, rotateDegree),
    );
  }

  @override
  void dispose() {
    super.dispose();
    timer.cancel();
  }

  void _initLeavesColors() {
    Color startColor = widget.color;
    Color endColor = Color(widget.color.value & 0x00FFFFFF | 0x11000000);
    for (int i = 0; i < widget.leavesCount; i++) {
      double fraction = 1 - i * 30 / 360;
      widget.leaveColors.add(_evaluate(fraction, startColor, endColor));
    }
  }

  Color _evaluate(double fraction, Color startValue, Color endValue) {
    int startInt = startValue.value;
    double startA = ((startInt >> 24) & 0xff) / 255.0;
    double startR = ((startInt >> 16) & 0xff) / 255.0;
    double startG = ((startInt >> 8) & 0xff) / 255.0;
    double startB = (startInt & 0xff) / 255.0;
    int endInt = endValue.value;
    double endA = ((endInt >> 24) & 0xff) / 255.0;
    double endR = ((endInt >> 16) & 0xff) / 255.0;
    double endG = ((endInt >> 8) & 0xff) / 255.0;
    double endB = (endInt & 0xff) / 255.0;
    // convert from sRGB to linear
    startR = Math.pow(startR, 2.2);
    startG = Math.pow(startG, 2.2);
    startB = Math.pow(startB, 2.2);
    endR = Math.pow(endR, 2.2);
    endG = Math.pow(endG, 2.2);
    endB = Math.pow(endB, 2.2);
    // compute the interpolated color in linear space
    double a = startA + fraction * (endA - startA);
    double r = startR + fraction * (endR - startR);
    double g = startG + fraction * (endG - startG);
    double b = startB + fraction * (endB - startB);
    // convert back to sRGB in the [0..255] range
    a = a * 255.0;
    r = Math.pow(r, 1.0 / 2.2) * 255.0;
    g = Math.pow(g, 1.0 / 2.2) * 255.0;
    b = Math.pow(b, 1.0 / 2.2) * 255.0;
    return Color(
        a.round() << 24 | r.round() << 16 | g.round() << 8 | b.round());
  }
}

class _IndicatorPainter extends CustomPainter {
  final Paint p;
  final Color color;
  final leavesCount;
  final leaveColors;
  final Path path;
  final double rotateDegree;

  _IndicatorPainter(this.p,
                    this.color,
                    this.leavesCount,
                    this.leaveColors,
                    this.path,
                    this.rotateDegree,);

  @override
  void paint(Canvas canvas, Size size) {
    _drawFlower(canvas, size);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }

  void _drawFlower(Canvas canvas, Size size) {
    canvas.translate(size.width / 2, size.width / 2);
    canvas.rotate(rotateDegree * Math.pi / 180);
    canvas.save();
    double leafWidth = size.width / 12;
    path.reset();
    Rect rect = Rect.fromLTRB(
        -leafWidth / 2, -size.width / 2, leafWidth / 2, -size.width / 4);
    RRect rRect = RRect.fromRectAndRadius(rect, Radius.circular(leafWidth / 2));
    path.addRRect(rRect);
    path.close();
    for (int i = 0; i < leavesCount; i++) {
      Color color = leaveColors[i];
      p.color = color;
      canvas.drawPath(path, p);
      canvas.rotate(-30 * Math.pi / 180);
    }
    canvas.restore();
  }
}


example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:loading_hud/loading_hud.dart';
import 'package:loading_hud/loading_indicator.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'LoadingHUD',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('LoadingHUD'),
      ),
      body: ListView(
        children: <Widget>[
          _buildListItem(
            context,
            title: 'Show default',
            onTap: () => _buildDefaultHud(context).show(),
          ),
          _buildListItem(
            context,
            title: 'Show for 3 seconds',
            onTap: () {
              var progressHud = _buildUnCancelableHud(context);
              progressHud.show();
              Future.delayed(Duration(seconds: 3), () {
                progressHud.dismiss();
              });
            },
          ),
          _buildListItem(
            context,
            title: 'Show with background dimless',
            onTap: () => _buildNoDimBackgroundHud(context).show(),
          ),
          _buildListItem(
            context,
            title: 'Show success',
            onTap: () => _buildSuccessHud(context).show(),
          ),
          _buildListItem(
            context,
            title: 'Show error',
            onTap: () => _buildErrorHud(context).show(),
          ),
        ],
      ),
    );
  }

  Widget _buildListItem(BuildContext context,
      {String title, VoidCallback onTap}) {
    return Container(
      decoration: BoxDecoration(
        border: Border(
          bottom: BorderSide(
            color: Colors.grey,
            width: 0,
          ),
        ),
      ),
      child: ListTile(
        title: Text(title),
        onTap: onTap,
      ),
    );
  }

  LoadingHud _buildDefaultHud(BuildContext context) {
    return LoadingHud(context);
  }

  LoadingHud _buildUnCancelableHud(BuildContext context) {
    return LoadingHud(
      context,
      cancelable: false,
      canceledOnTouchOutside: false,
    );
  }

  LoadingHud _buildNoDimBackgroundHud(BuildContext context) {
    return LoadingHud(
      context,
      dimBackground: false,
      hudColor: Color(0x99000000),
      indicator: DefaultLoadingIndicator(
        color: Color(0xFFFFFFFF),
      ),
    );
  }

  LoadingHud _buildSuccessHud(BuildContext context) {
    return LoadingHud(
      context,
      iconSuccess: Icon(Icons.done),
      iconError: Icon(Icons.error),
      future: onSuccess(),
    );
  }

  Future<Text> onSuccess() {
    return Future.delayed(Duration(seconds: 2), () {
      return Text('That was great!');
    });
  }

  LoadingHud _buildErrorHud(BuildContext context) {
    return LoadingHud(
      context,
      hudColor: Colors.black54,
      indicator: DefaultLoadingIndicator(
        color: Color(0xFFFFFFFF),
      ),
      iconSuccess: Icon(
        Icons.done,
        color: Colors.white,
      ),
      iconError: Icon(
        Icons.error,
        color: Colors.white,
      ),
      future: onError(),
    );
  }

  Future<Text> onError() {
    try {
      return Future.delayed(Duration(seconds: 2), () {
            throw Text(
              'Something went wrong.',
              style: TextStyle(
                fontSize: 16,
                color: Colors.white,
                decoration: TextDecoration.none,
              ),
            );
          });
    } catch (e) {
      print(e);
    }
  }
}
